### ISSUE TITLE
'Add column(s) from other OpenRefine project' as default feature

### ISSUE THREAD

**Issue by trnstlntk** [OpenRefine/OpenRefine#5393]
I have the impression that [joining projects](https://guides.library.illinois.edu/openrefine/joiningprojects) (adding one or more columns from other OpenRefine projects based on a specified key column) is a very common use case. I do it a lot myself and have the impression - judging from the many explainers and tutorials online and the consistent feedback in OpenRefine's user survey - that many users also frequently do it.

Whenever I need to do it, I always end up Googling how it should be done and then doing the puzzle of constructing the intricate `cell.cross('arg1','arg2').cells['arg3'].value[arg4]` construct; writing down the exact project name, column names + debugging when I inevitably misspelled something. It's not the most pleasant thing to do.

### Proposed solution
Add a default feature to OpenRefine with a dialog window that allows users to do it without having to look up the GREL function.

### Additional context
I heard that the VIB-BITS extension has this feature.

**Comment by trnstlntk** [comment]
Some considerations:

* I am not a designer and have no strong opinions on how this should be built.
* I would personally not be inclined to install an extension (the VIB-BITS one) specifically for this. How well maintained is that extension? How will most users even know that this extension even exists? IMO this feature is so common that there are good arguments for having it by default in OpenRefine.

**Comment by wetneb** [comment]
Absolutely! I think it would still be worth trying out the VIB-BITS extension so that you can assess the user experience it provides, and indicate what you would change there.

CC-ing @swkasica who has been thinking about this

**Comment by tfmorris** [comment]
The VIB-BITS extension is GPL licensed, so, unfortunately, its code can't be used as a starting point unless they are willing to relicense it.

re:
> How well maintained is that extension?

It hasn't been touched since 2015, so not very well.

**Comment by swkasica** [comment]
I agree that this is a common data preparation task. Do folks know if VIB-BITS is the only extension that has features similar to relational join operations?

**Comment by ostephens** [comment]
As far as I'm aware in terms of adding columns, the VIB-BITS extension is just a wrapper for the existing 'cross' function and related operations - so its a UI over existing functionality rather than adding any support for join type operations

**Comment by thadguidry** [comment]
One thing we still don't do well is making it easier to select not only 1 key, but multiple keys between 2 datasets (currently, you have to manually concatenate columns/keys into a single column, in order to do so) and then performing an inner join or even a union to combine the 2 datasets.  I never wanted database joins and unions to stump users with us pushing more into some sort of query language syntax, but instead always wanted our users to have an intuitive muscle-memory joining dialog to help navigate this complex area.  It's also something that Google and ProPublica wanted at one time and still do.  And at one point in time, David and I even discussed a better UX to support more than 2 datasets through multiway merge (which Pentaho had as a multiway merge join step).

**Comment by wumionwheels** [comment]
Hello, I am an outreachy applicant. I am interested in this and I have two proposals
1. Is to add an add column from other projects button(can be simplified to add columns from) before the create project button
2.  is to add an add columns button after project creation beside the open, export and help buttons

**Comment by wumionwheels** [comment]
These are mock-ups of my proposal
1. Is to add an add column from other projects button(can be simplified to add columns from) before the create project button. I also used the chance to add a scrolling feature to Parse data as card, this will let users scroll the card without scrolling the entire bottom part of the page thereby preventing other information on the bottom part of the page that could be static from scrolling up
<img width="1440" alt="Openrefine add column button 1" src="https://user-images.githubusercontent.com/23156958/224034493-3d74b39c-64c1-49c3-bca4-817cbadba017.png">

2. Is to add an add column from other projects button(can be simplified to add columns from) before the create project button
<img width="1440" alt="Openrefine Add column button 2" src="https://user-images.githubusercontent.com/23156958/224034711-f55a5a60-11d0-4301-b214-a098cc1bcee3.png">

**Comment by wetneb** [comment]
I think this feature should only exist in the project view, not the preview where parsing options can be changed.

In the project view, beyond the question of how the feature should be integrated in the existing menus and tool bars, I think the main question is what would actually happen when you click your "Add Column(s)…" button. I assume you would get some dialog to specify some parameters? What would it look like? And in what state would the project grid end up after you validate this dialog?

**Comment by lozanaross** [comment]
@wumionwheels - do you want to spend a bit more time on this issue and investigate it properly? if yes, I'll assign it to you and you can then take your time before you submit complete proposals to us including the dialog that @wetneb mentions.

**Comment by wumionwheels** [comment]
OK, I will do that.

**Comment by wumionwheels** [comment]
After taking @wetneb feedback into account, this is the result. Hoping to receive more feedback to further iterate on

### Add column(s) flow Variant 1 

**Click on add column(s) button**

<img width="1440" alt="Openrefine Add column before (1)" src="https://user-images.githubusercontent.com/23156958/224563803-deea05d3-a693-4757-af77-8744c7c77458.png">



**A dialogue overlay opens up asking users to select projects**

<img width="1440" alt="Openrefine Add column dailogue" src="https://user-images.githubusercontent.com/23156958/224563064-496315a1-1a97-4ba8-8bd1-8f6d893957dd.png">



**When user clicks on project input field, their openrefine projects are listed and can also be searched for faster result**

<img width="1440" alt="Openrefine Add project open" src="https://user-images.githubusercontent.com/23156958/224563124-e2e56e2e-3690-4aee-92cb-ce5b1639b825.png">



**Upon selecting a project, an add column button is displayed**

<img width="1440" alt="Openrefine Add project selected and add column" src="https://user-images.githubusercontent.com/23156958/224563158-4b36437f-7d2b-46fb-a344-bedf846fe222.png">



**The add column input field with a drop-down functions like the select project input field, columns in the project are listed and can be searched for faster result.
users can also add another column from the chosen project and can add another project**

<img width="1440" alt="Openrefine Add coumn variant 1" src="https://user-images.githubusercontent.com/23156958/224563455-b34076c9-6ec3-4bde-b454-8d0b2306e0ed.png">



**How the dialogue will look when another column is added**

<img width="1440" alt="Openrefine Add column 2columns variant 1" src="https://user-images.githubusercontent.com/23156958/224563467-e5e5af55-8506-43f9-8230-011d6b3ef5e5.png">



**How the dialogue will look when another project is added**

<img width="1440" alt="Openrefine Add another project variant 1" src="https://user-images.githubusercontent.com/23156958/224563501-4c1e577b-a83d-4184-a4b5-09ae1759c353.png">



**How the dialogue will look when it is scrolled up**

<img width="1440" alt="Openrefine Add another project scrolled up variant 1" src="https://user-images.githubusercontent.com/23156958/224563514-6a962f3c-7019-4c5b-ac0c-39deed0e6754.png">



**Here I considered allowing user to choose the position of their added column(s), though I could only have two positions. I also considered using blue text buttons instead of the radio button used but decided against it because of how blue text button is used in the dialogue(in the case of add column and button) so as not to confuse users.
another solution is to just automatically add column(s) to the beginning**

<img width="1440" alt="Openrefine Add coumn position for variant 1" src="https://user-images.githubusercontent.com/23156958/224563525-c06bf9be-5a80-4889-a5a4-65520b17cf67.png">



**Final project page outlook with an alert of added column(s) position**

<img width="1440" alt="Openrefine Add column After (1)" src="https://user-images.githubusercontent.com/23156958/224563788-d430c7ab-bf8c-42b0-a7ee-e83078c8b185.png">



video of prototype for flow variant 1

[New Recording - 12_03_2023, 18_47_46.webm](https://user-images.githubusercontent.com/23156958/224563880-8b2aab73-2ddb-4165-aca2-9a3197392256.webm)



### Add column(s) flow Variant 2
the only difference is where add another column button is placed

<img width="8995" alt="Add column(s) flow Variant 2" src="https://user-images.githubusercontent.com/23156958/224564759-26cc52a0-0800-4200-80b6-afd169d7c19c.png">



### Add column(s) flow Variant 3
the difference is where add another column button is placed, and it is represented by a plus icon

<img width="8995" alt="Add column(s) flow Variant 3" src="https://user-images.githubusercontent.com/23156958/224564783-32a61a6f-0bb6-4cef-b2e3-5ef2466acec7.png">

**Comment by thadguidry** [comment]
@wumionwheels Great work thus far!
A few things from me that I'd like to see to polish this up for consistency with OpenRefine's other existing dialogs.

* Add a background header sized and with color `#E0EDFE` to the top of the dialog (look at Edit Cells -> Transform, as an example)
* Add a footer sized and with color `#EEE` for the Cancel | Add buttons.
* @wetneb @lozanaross Not sure what to do about the Cancel | Add ordering... did we decide in other issues to have Cancel always on left most button?  I've already forgotten.
* I'd likely want to see a faint bordering effect around each Project level to box it off and show the separation of each.  We have bordering box examples in many places such as the Edit Cells -> Transform, with it's Preview grid - `.ui-widget.ui-widget-content` with color `#CCCCCC`

**Comment by wetneb** [comment]
Before getting into visual details I think it would be really good to step back a bit and think about what the requested feature is really about. As stated in the first message this is about *joining* projects, i.e. offering an interface which simplifies the use of the `cross` function. This is not about simply copying one project to another (which would probably only make sense if the projects have the same number of rows and would generally not be very useful in comparison).

To join projects it is not sufficient to pick another column in another project. One also needs to pick which columns (of the current project and of the target project) should be used as common keys.

So, to add a single column, this dialog must get (at least) four pieces of information from the user:
* the target project
* the column to use in the target project as key
* the corresponding key column in the current project
* the column(s) to fetch in the target project

Again, as mentioned above, I would recommend to check out the UX of the VIB-BITS plugin which does something similar. That would avoid misunderstandings.

**Comment by wumionwheels** [comment]
Ok thank you so much @wetneb I will make sure to go through the vib-bits plugin and not rush into designing before I get it, please will it be okay perhaps before designing to post my flow here so that you will give feedback before I start designing?

@thadguidry thank you so much for your feedback, I will work on including these when I am in the design phase again.

**Comment by wumionwheels** [comment]
@wetneb just to know if my research is going in the right direction when you said key column, do you mean matching column?

**Comment by wumionwheels** [comment]
I saw a document about how VIB-BITS Plugin join feature works, but I don't understand the aggregation method part, will be glad if someone can shed more light on it
![FireShot Capture 011 - H Appendix_ VIB-BITS extension - OpenRefine 3 4 tutorial_ cleaning, p_ - msaby gitlab io](https://user-images.githubusercontent.com/23156958/224684939-fa1b1ceb-190b-46df-aeaa-aad216371f9c.png)

![FireShot Capture 008 - H Appendix_ VIB-BITS extension - OpenRefine 3 4 tutorial_ cleaning, p_ - msaby gitlab io](https://user-images.githubusercontent.com/23156958/224684846-e04e86ff-654a-444f-8d78-e7a285d223da.png)

This is the link for reference
https://msaby.gitlab.io/tutoriel-openrefine/annexe-extension-vib-bits.html

**Comment by wetneb** [comment]
True, it looks like the homepage for the extension is down. Sad! But you should be able to download the VIB-BITS plugin via the Wayback Machine:
http://web.archive.org/web/20220307183211/https://data.bits.vib.be/hidden/g7dt6RjuUTU421dY2CwrGePGX/vib-bits.zip

**Comment by wumionwheels** [comment]
ok thanks

**Comment by wumionwheels** [comment]
@trnstlntk Please for the arg4 funtion which tells what matching value to choose in case of multiple matches
arg4 = indicate which value to import in the array (if multiple matches for the key) (recommended to use 0)
is the key here referring to the matching column(key column) of the project from which data will be exported from or the columns that will be exported

**Comment by trnstlntk** [comment]
> @trnstlntk Please for the arg4 funtion which tells what matching value to choose in case of multiple matches
> 
> arg4 = indicate which value to import in the array (if multiple matches for the key) (recommended to use 0)
> 
> is the key here referring to the matching column(key column) of the project from which data will be exported from or the columns that will be exported 
> 
> 

I actually don't know 😂 maybe others in this thread do?

**Comment by wumionwheels** [comment]
😂 Alright, thanks. I will continue to research it

**Comment by wumionwheels** [comment]
@trnstlntk upon doing further research, it is referring to the key column

**Comment by wumionwheels** [comment]
@wetneb please correct me if I am wrong, from my research I discovered that the key column is a column in a table that contains unique values that can be used to identify and attribute data in other columns and to join projects together we have to identify the key column in the two project and they have to match but from my interactions with VIB-BITS Plugin any column can be used as a key column it just have to match, so is it safe to say that any column can be the key column since all columns have values that can be used to identify and attribute data

**Comment by wetneb** [comment]
That's correct. Note that you could even use the row numbers as keys. Row numbers do not form a column in OpenRefine so that might require tweaks in the UI to let the user choose them if we want to support that. I do not know if the VIB-BITS plugin supports that but it is at least something that can be done with the underlying `cross` function.

**Comment by ostephens** [comment]
@wetneb said:

> So, to add a single column, this dialog must get (at least) four pieces of information from the user:
> 
> the target project
> the column to use in the target project as key
> the corresponding key column in the current project
> the column(s) to fetch in the target project

Just to note that usually (in the current OpenRefine UX) a function like "Add column" is triggered from the drop down menu at the top of a column. This is where the current "Add column" options exist
 
<img width="668" alt="Screenshot 2023-03-14 at 12 04 00" src="https://user-images.githubusercontent.com/576174/224995596-4efa9db1-88ab-4e38-b4f8-041e0016b19a.png">

Using this location for the add column from another project option would be both consistent with current UX and also mean that the "the corresponding key column in the current project" mentioned by @wetneb is defined automatically - it is the one the user is currently working from. This is also consistent with the current UX

**Comment by ostephens** [comment]
@wumionwheels you asked about:
> but I don't understand the aggregation method part [in the VIB-BITs extension], will be glad if someone can shed more light on it

and

> Please for the arg4 funtion which tells what matching value to choose in case of multiple matches
arg4 = indicate which value to import in the array (if multiple matches for the key) (recommended to use 0)
is the key here referring to the matching column(key column) of the project from which data will be exported from or the columns that will be exported

I hope I can explain a little.

The VIB-BITs extension is making use of an existing functionality in OpenRefine called `cross` to get data from other projects. The `cross` function takes a value from one project, and looks for it in a specified column in a second project. For each time it finds the value in the column the second project the `cross` function grabs the entire row from the second project

So project one:
|RowIndex|Key column|
|----------|-----------|
| 0 | KeyOne |
| 1 | KeyTwo |
| 2 | KeyThree |

And project 2:
|RowIndex|Target column|Column1|Column2|
|----------|---------------|---------|---------|
| 0 |KeyOne| ValueOne | ValueA |
| 1 |KeyTwo| ValueTwo | ValueB |
| 2 |KeyTwo | ValueThree | Value C |

If we do a `cross` from project one from the `Key column` to project two `Target column` then we have the following cases:

### Case 1
For the first row (value 'KeyOne') we find a single match in project two Target column and get back the entire row:
|RowIndex|Target column|Column1|Column2|
|----------|---------------|---------|---------|
| 0 |KeyOne| ValueOne | ValueA |

### Case 2
For the second row (value 'KeyTwo') we find two matches in project two Target column and so get back two rows:
|RowIndex|Target column|Column1|Column2|
|----------|---------------|---------|---------|
| 1 |KeyTwo| ValueTwo | ValueB |
| 2 |KeyTwo | ValueThree | Value C |

### Case 3
And finally, for the third row (value 'KeyThree') we find no matches and get back no rows.


So the `cross` function results in one of:
* `null` (no matches found = Case 3)
* an error (not illustrated here but included for completeness)
* an array of rows from the second project containing one or more rows (Case 1 and Case 2)

In the case that there is a single row returned (Case 1) then in terms of the data we can extract we are limited to just that single row (and we can extract values from one or more columns in that row). For example, we could extract the value from Column1 which in Case 1 would be "ValueOne"

However in the case that multiple rows are returned (Case 2) then for each column we want to extract into our project we have multiple values to deal with. So for example, if we have Case 2 and want to get the value from Column1 then we would get two values "ValueTwo" and "ValueThree". So we need decide what to do with this list of values that we've got.

This is what the aggregation method in the VIB-BITs extension is for. It offers the options:
* First (take the first value from all available values returned from the `cross`)
* Last (take the first value from all available values returned from the `cross`)
* Concatenate (join together all the values returned from the `cross` into a single string)
* Unique concatenate (join together all the values returned from the `cross` into a single string but ensuring each value appears only once)

I've given a lot of detail here which I hope is helpful but I realise it's a lot! Please ask if I can clarify further

**Comment by wumionwheels** [comment]
@wetneb regarding using row numbers, I think it can be included if it won't confuse the users

**Comment by wumionwheels** [comment]
@ostephens upon trying the first, last, concatenate and unique concatenate, I understand their purpose more, but the split afterwards is not working which makes me to wonder what its use is

**Comment by wumionwheels** [comment]
@wetneb I take back my back my statement that using row numbers may be confusing for the user especially after working on adding a data with multiple key matches. when I use first, it automatically adopt using the first value that matches for all the values. From the example shown below, the importance of have row numbers show in the UI as a column can be seen

**Table 1**
| Customer Name  | Order Quantity |
| ------------- | ------------- |
| Jane James  | 40  |
| Lara Luke  | 34  |
| Jane James  | 4  |
| Lara Luke  | 3  |
| Jane James  | 13  |
| Lara Luke  | 4  |


**Table 2**
| Customer Name  | Order Quantity |
| ------------- | ------------- |
| Jane James  | 40  |
| Lara Luke  | 34  |
| Jane James  | 49  |
| Lara Luke  | 30  |
| Jane James  | 130  |
| Lara Luke  | 4  |


when I add order quantity column from Table 2 to Table 1 based on Customer name column (key column) this is the result if 'arg4' value (aggregation method in VIB-BITS Plugin) is;

1. First (0)
**Table 1**

 | Customer Name | Order Quantity2 | Order Quantity |
 | ------------- | ------------- | ------------- |
 | Jane James | 40  | 40  |
 | Lara Luke | 34  | 34  |
| Jane James | 40  | 4  |
| Lara Luke | 34  | 3  |
| Jane James | 40  | 13  |
| Lara Luke | 34  | 4  |

2. Last  (2)
**Table 1**

 | Customer Name | Order Quantity2 | Order Quantity |
 | ------------- | ------------- | ------------- |
 | Jane James | 13  | 40  |
 | Lara Luke | 4  | 34  |
| Jane James | 13  | 4  |
| Lara Luke | 4  | 3  |
| Jane James | 13  | 13  |
| Lara Luke | 4  | 4  |

3. Concatenate (joining them together) using ',' as separator 
**Table 1**

 | Customer Name | Order Quantity2 | Order Quantity |
 | ------------- | ------------- | ------------- |
 | Jane James | 40,4,13  | 40  |
 | Lara Luke | 34,3,4  | 34  |
| Jane James | 40,4,13  | 4  |
| Lara Luke | 34,3,4  | 3  |
| Jane James | 40,4,13  | 13  |
| Lara Luke | 34,3,4  | 4  |

4. As for unique concatenate, it will be the same as 3 since we don't value in order quantity column for Jane James and Lara Luke that are the same

**Comment by wumionwheels** [comment]
oh @ostephens  thanks, I will check it out

**Comment by wumionwheels** [comment]
@wetneb regarding my recent message, we can have the key column of the current project on the dialogue, it will be in form of a selected radio button to show it is already selected and there will be a change text button beside it so that users can change it if they want. Just that I am not sure how the dev part will work

**Comment by wumionwheels** [comment]
@lozanaross @wetneb @ostephens @trnstlntk and @thadguidry  please your feedbacks are needed

**Comment by wetneb** [comment]
Hi @wumionwheels, at a quick glance this looks much better but please bear in mind that we have a lot on our plate. There is no need to ping us 16 hours after your last message. Could you please wait a bit more? For feedback on such a big proposal I would wait for a few days at least. Whether or not this is considered a final proposal, I am sure you can already register it in Outreachy as a contribution. Thanks a lot for your work and your patience!

**Comment by wumionwheels** [comment]
oh I am so sorry for that, I will delete the message

**Comment by trnstlntk** [comment]
Thanks @wumionwheels for this proposal!

I don't know what the plans are for the Outreachy design internship, but @lozanaross IMO it would be super valuable to do some varied user tests for larger proposals like this? As you have done for the Commons project: ask the Outreachy design intern to place this design in front of a variety of OpenRefine users (both less and more experienced) and improve it based on their feedback? Just a suggestion :-)

**Comment by wumionwheels** [comment]
I will surely love to work on testing this after receiving a cue from @lozanaross, thank you so much @trnstlntk for your valuable suggestion

**Comment by lozanaross** [comment]
@wumionwheels Thanks for your work so far - this is indeed a large proposal and you can already record it as a contribution. Whether this is the final version that gets built will need to involve further user testing that is best done in the context of the actual internship. Since we haven't selected the final candidate for the internship yet, we can't make concrete plans for next steps. 

As quick design feedback (that is based on personal experience and doesn't preclude the need for user testing), I would suggest we stick with the proposal made by @ostephens that adding new columns is done via the existing UX pattern (instead of adding new buttons in the top): 

> "Just to note that usually (in the current OpenRefine UX) a function like "Add column" is triggered from the drop down menu at the top of a column. This is where the current "Add column" options exist
> 
> Using this location for the add column from another project option would be both consistent with current UX and also mean that the "the corresponding key column in the current project" mentioned by @wetneb is defined automatically - it is the one the user is currently working from. This is also consistent with the current UX"

In this case, we don't need to specify (or offer option to change) current key column (i.e. the first bit in your dialog design) - this will have the added benefit of reducing complexity of what is already a pretty complex dialog. When testing we should also test if the term "key column" is understandable to users. We also need to offer an option to use the row numbers as keys. This needs a special radio button or tick box I think - but again, a feature to be tested further.

Otherwise, I think I prefer design option 2 from above which also features the Preview mode - this is consistent with other dialogs in OpenRefine and is useful to the users - to see if they have selected what they want indeed.

**Comment by wumionwheels** [comment]
OK @lozanaross, thank you so much for your feedback
